include/deal.II/lac/la_sm_partitioner.h

```diff
@@ -0,0 +1,482 @@
+
+      /**
+       * Finish to import from ghost array.
+       */
+      virtual void
+      import_from_ghosted_array_finish(
+        const VectorOperation::values        operation,
+        double *const                        data_this,
+        const std::vector<double *> &        data_others,
+        const dealii::AlignedVector<double> &buffer,
+        std::vector<MPI_Request> &           requests) const = 0;
+
+      /**
+       * Start to export to ghost array.
+       */
+      virtual void
+      export_to_ghosted_array_start(
+        const unsigned int            communication_channel,
+        float *const                  data_this,
+        const std::vector<float *> &  data_others,
+        dealii::AlignedVector<float> &buffer,
+        std::vector<MPI_Request> &    requests) const = 0;
+
+      /**
+       * Finish to export to ghost array.
+       */
+      virtual void
+      export_to_ghosted_array_finish(
+        float *const                data_this,
+        const std::vector<float *> &data_others,
+        std::vector<MPI_Request> &  requests) const = 0;
+
+      /**
+       * Start to import from ghost array.
+       */
+      virtual void
+      import_from_ghosted_array_start(
+        const VectorOperation::values operation,
+        const unsigned int            communication_channel,
+        float *const                  data_this,
+        const std::vector<float *> &  data_others,
+        dealii::AlignedVector<float> &buffer,
+        std::vector<MPI_Request> &    requests) const = 0;
+
+      /**
+       * Finish to import from ghost array.
+       */
+      virtual void
+      import_from_ghosted_array_finish(
+        const VectorOperation::values       operation,
+        float *const                        data_this,
+        const std::vector<float *> &        data_others,
+        const dealii::AlignedVector<float> &buffer,
+        std::vector<MPI_Request> &          requests) const = 0;
+
+    private:
+      /**
+       * Flag indicating if the partitioner implementation has been enabled
+       * with contiguous or non-noncontiguous shared-memory allocation.
+       */
+      bool contiguous_allocation;
+
+    protected:
+      /**
+       * Global communicator.
+       */
+      MPI_Comm comm;
+
+      /**
+       * Shared-memory sub-communicator.
+       */
+      MPI_Comm comm_sm;
+
+      /**
+       * Number of processes in comm.
+       */
+      unsigned int n_mpi_processes_;
+
+      /**
+       * Number of locally-owned vector entries.
+       */
+      unsigned int n_local_elements;
+
+      /**
+       * Number of ghost vector entries.
+       */
+      unsigned int n_ghost_elements;
+    };
+
+    /**
+     * Partitioner implementation that is built around index-sets similar to
+     * Utilities::MPI::Partitioner, however, distinguishes between remote and
+     * shared ghost vector entries.
+     */
+    class Partitioner : public PartitionerBase
```

Comment:
The members are part of the base class because they are needed by some methods in the base class. I don't think these methods need the be implemented by the implementations.

Reply:
OK.